drivers/espi/espi_realtek_rts5912.c

```diff
@@ -0,0 +1,236 @@
+	const struct device *clk_dev;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct espi_rts5912_data {
+	sys_slist_t callbacks;
+	uint32_t config_data;
+};
+
+/*
+ * =========================================================================
+ * ESPI common function and API
+ * =========================================================================
+ */
+
+#define RTS5912_ESPI_MAX_FREQ_20 20
+#define RTS5912_ESPI_MAX_FREQ_25 25
+#define RTS5912_ESPI_MAX_FREQ_33 33
+#define RTS5912_ESPI_MAX_FREQ_50 50
+#define RTS5912_ESPI_MAX_FREQ_66 66
+
+static int espi_rts5912_configure(const struct device *dev, struct espi_cfg *cfg)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	struct espi_rts5912_data *data = (struct espi_rts5912_data *)dev->data;
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+
+	uint32_t gen_conf = 0;
+	uint8_t io_mode = 0;
+
+	/* Maximum Frequency Supported */
+	switch (cfg->max_freq) {
+	case RTS5912_ESPI_MAX_FREQ_20:
+		gen_conf |= 0UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_25:
+		gen_conf |= 1UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_33:
+		gen_conf |= 2UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_50:
+		gen_conf |= 3UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_66:
+		gen_conf |= 4UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* I/O Mode Supported */
+	io_mode = cfg->io_caps >> 1;
+
+	if (io_mode > 3) {
+		return -EINVAL;
+	}
+	gen_conf |= io_mode << ESPI_ESPICFG_IOSUP_Pos;
+
+	/* Channel Supported */
+	if (cfg->channel_caps & ESPI_CHANNEL_PERIPHERAL) {
+		gen_conf |= BIT(0) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_VWIRE) {
+		gen_conf |= BIT(1) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_OOB) {
+		gen_conf |= BIT(2) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_FLASH) {
+		gen_conf |= BIT(3) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	espi_reg->ESPICFG = gen_conf;
+	data->config_data = espi_reg->ESPICFG;
+
+	return 0;
+}
+
+static bool espi_rts5912_channel_ready(const struct device *dev, enum espi_channel ch)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+
+	switch (ch) {
+	case ESPI_CHANNEL_PERIPHERAL:
+		return espi_reg->EPCFG & ESPI_EPCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_VWIRE:
+		return espi_reg->EVCFG & ESPI_EVCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_OOB:
+		return espi_reg->EOCFG & ESPI_EOCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_FLASH:
+		return espi_reg->EFCONF & ESPI_EFCONF_CHEN ? true : false;
+	default:
+		return false;
+	}
+}
+
+static int espi_rts5912_manage_callback(const struct device *dev, struct espi_callback *callback,
+					bool set)
+{
+	struct espi_rts5912_data *data = (struct espi_rts5912_data *)dev->data;
```

Comment:
Cast isn't needed here or in any of the other routines.
```suggestion
	struct espi_rts5912_data *data = dev->data;
```